_posts/2018-11-19-bad-blocks.md

```diff
@@ -0,0 +1,147 @@
+Currently (Linux kernel v4.19, libndctl v62) these operations require
+read access to the following resource files (containing physical addresses)
+of NVDIMM devices which only the super user can read by default:
+```
+	/sys/bus/nd/devices/ndbus*/region*/resource
+	/sys/bus/nd/devices/ndbus*/region*/dax*/resource
+	/sys/bus/nd/devices/ndbus*/region*/pfn*/resource
+	/sys/bus/nd/devices/ndbus*/region*/namespace*/resource
+```
+
+
+### Opening and creating a pool
+
+If the `CHECK_BAD_BLOCKS` compat feature is enabled, every pool is checked
+if it contains bad blocks during opening and during creating
+(when the pool is created using an already existing zeroed file(s)).
+If it does then opening/creating fails.
+
+If the `CHECK_BAD_BLOCKS` compat feature is enabled and the user does not have
+enough permissions (see 
+[pmempool-feature](https://github.com/ldorau/pmem.github.io/blob/blog/_posts/2018-11-19-bad-blocks.md#pmempool-feature))
+to be able to check if the pool contains bad blocks then opening/creating
+fails either.
+
+
+### Pmempool-info
+
+If the `CHECK_BAD_BLOCKS` compat feature is enabled or ```--bad-blocks=yes```
+option is used then `pmempool-info` prints out information about bad blocks
+in the pool, for example:
+```
+$ pmempool info --bad-blocks=yes /dev/dax1.0
+Poolset structure:
+Number of replicas       : 1
+Replica 0 (master) - local, 1 part(s):
+part 0:
+path                     : /dev/dax1.0
+type                     : device dax
+size                     : 62922752
+alignment                : 4096
+bad blocks:
+        offset          length
+        11              1
+[...]
+```
+
+
+### Pmempool-check
+
+If the `CHECK_BAD_BLOCKS` compat feature is enabled, `pmempool-check` checks
+if the pool contains bad blocks, for example:
+```
+$ pmempool check /dev/dax1.0
+poolset contains bad blocks, use 'pmempool info --bad-blocks=yes' to print or 'pmempool sync --bad-blocks' to clear them
+/dev/dax1.0: cannot repair
+```
+
+
+### Pmempool-sync
+
+Attention: this feature is available **only for libpmemobj** pools.
+
+If the `CHECK_BAD_BLOCKS` compat feature is enabled or ```--bad-blocks```
+option is used then `pmempool-sync` tries to fix bad blocks in the libpmemobj
+pool using its replicas:
+
+```
+$ pmempool sync --bad-blocks /dev/dax1.0
+/dev/dax1.0: synchronized
+```
+
+Synchronization can fail if a part of the pool has uncorrectable errors
+in all replicas:
+
+```
+$ pmempool sync --bad-blocks /dev/dax1.0
+error: failed to synchronize: a part of the pool has uncorrectable errors in all replicas
+error: Invalid argument
+```
+
+Fixing bad blocks causes creating or reading special recovery files.
+When bad blocks are detected, special recovery files have to be created
+in order to fix them safely. A separate recovery file is created per each part
+of the pool. The recovery files are created in the same directory
+where the poolset file is located using the following name pattern:
```

Comment:
as the poolset file, using the following name pattern

Reply:
Done